### PR TITLE
feat(curation api): Update DataSubmission Lambda s3 object parsing.

### PR DESCRIPTION
**Pull request by Bento007**
- #TICKET_NUMBER

### Reviewers
**Functional:** 

**Readability:** 

---

## Changes
- Modify the expected format of the s3 object. Object must now conform to one of the following shapes:
  - S3://{bucket}/{user}/{collection}/{dataset}.h5ad or
  - S3://{bucket}/{user}/{collection}/{tag}.h5ad

### CODE REVIEW

File: backend/corpora/dataset_submissions/app.py [hunk 1]
```diff
@@ -6,14 +6,18 @@
...
+from backend.corpora.common.entities import Dataset
 from backend.corpora.common.upload import upload
 from backend.corpora.common.utils.db_session import db_session_manager
 from backend.corpora.common.utils.exceptions import CorporaException
 
 logger = logging.getLogger(__name__)
-USERNAME_REGEX = r"[\w\-\|]+"
+USERNAME_REGEX = r"(?P<username>[\w\-\|]+)"
 UUID_REGEX = r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+EXTENTION_REGEX = r"(?P<extension>h5ad)"
```
Comment (metakuni) [inline backend/corpora/dataset_submissions/app.py:17]:
```suggestion
EXTENSION_REGEX = r"(?P<extension>h5ad)"
```

File: backend/corpora/dataset_submissions/app.py [hunk 2]
```diff
@@ -6,14 +6,18 @@
...
+DATASET_ID_REGEX = f"(?P<dataset_uuid>{UUID_REGEX})"
+COLLECTION_ID_REGEX = f"(?P<collection_uuid>{UUID_REGEX})"
+REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((tag/(?P<tag>.*))|(id/{DATASET_ID_REGEX})).{EXTENTION_REGEX}$"
```
Comment (metakuni) [inline backend/corpora/dataset_submissions/app.py:20]:
```suggestion
REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((tag/(?P<tag>.*))|(id/{DATASET_ID_REGEX})).{EXTENSION_REGEX}$"
```

**Review by metakuni — APPROVED** [review]

**Comment by danieljhegeman** [comment]
This needs to be changed to expect `<bucket>/<collection_id>/<tag_or_dataset_id>` where `<tag_or_dataset_id>` is treated as an id (i.e., it is _necessarily_ replacing an existing dataset) if it conforms to uuid format and as a tag if it does not.

https://czi-sci.slack.com/archives/C02UGUV5R3J/p1652995597685589

**Review by danieljhegeman — CHANGES_REQUESTED** [review]
See the discussion about removing `/id/` and `/tag/` from the key, and instead relying on interpretation of the "tag" portion.

**Comment by Bento007** [comment]
@danieljhegeman [resolved](https://czi-sci.slack.com/archives/C02UGUV5R3J/p1652995597685589)

File: tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py [hunk 3]
```diff
@@ -20,30 +20,25 @@ def _test_missing_fields(self, **kwargs):
...
-
-    def test__missing_curator_id__raises_error(self):
-        self._test_missing_fields(key=f"{self.user_name}/{self.collection_uuid}/id/")
-
-    def test__missing_type__raise_error(self):
-        self._test_missing_fields(key=f"{self.user_name}/{self.collection_uuid}/{self.incoming_curator_tag}")
+    def test__missing_curator_file_name__raises_error(self):
+        self._test_missing_fields(key=f"{self.user_name}/{self.collection_uuid}/")
 
     def test__misssing_extension__raise_error(self):
```
Comment (metakuni) [inline tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py:26]:
```suggestion
    def test__missing_extension__raise_error(self):
```

File: tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py [hunk 4]
```diff
@@ -20,30 +20,25 @@ def _test_missing_fields(self, **kwargs):
...
-        self._test_missing_fields(key=f"{self.user_name}/{self.collection_uuid}/id/{self.dataset_uuid}")
+        self._test_missing_fields(key=f"{self.user_name}/{self.collection_uuid}/{self.dataset_uuid}")
 
     def test__missing_collection_uuid__raises_error(self):
-        self._test_missing_fields(key=f"{self.user_name}/should_have_been_a_uuid/tag/{self.incoming_curator_tag}")
+        self._test_missing_fields(key=f"{self.user_name}/should_have_been_a_uuid/t{self.incoming_curator_tag}")
```
Comment (metakuni) [inline tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py:30]:
```suggestion
        self._test_missing_fields(key=f"{self.user_name}/should_have_been_a_uuid/{self.incoming_curator_tag}")
```

**Review by metakuni — APPROVED** [review]

File: backend/corpora/dataset_submissions/app.py [hunk 5]
```diff
@@ -6,14 +6,18 @@
...
 
-from backend.corpora.common.entities import Collection
+from backend.corpora.common.entities import Dataset
 from backend.corpora.common.upload import upload
 from backend.corpora.common.utils.db_session import db_session_manager
 from backend.corpora.common.utils.exceptions import CorporaException
 
 logger = logging.getLogger(__name__)
-USERNAME_REGEX = r"[\w\-\|]+"
+USERNAME_REGEX = r"(?P<username>[\w\-\|]+)"
```
Comment (danieljhegeman) [inline backend/corpora/dataset_submissions/app.py:15]: cooool 🧊 

I did not know about 'named capture groups' in python. Glad to be introduced. 🧠

File: backend/corpora/dataset_submissions/app.py [hunk 6]
```diff
@@ -6,14 +6,18 @@
...
 UUID_REGEX = r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+EXTENSION_REGEX = r"(?P<extension>h5ad)"
+DATASET_ID_REGEX = f"(?P<dataset_uuid>{UUID_REGEX})"
+COLLECTION_ID_REGEX = f"(?P<collection_uuid>{UUID_REGEX})"
+REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((?P<tag>.*)|{DATASET_ID_REGEX}).{EXTENSION_REGEX}$"
```
Comment (danieljhegeman) [inline backend/corpora/dataset_submissions/app.py:20]: gorgeous 🤩

Comment (danieljhegeman) [reply backend/corpora/dataset_submissions/app.py:20]: But one problem. suggested fix:

```suggestion
REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((?P<tag>.*)|{DATASET_ID_REGEX})\\.{EXTENSION_REGEX}$"
```

File: backend/corpora/dataset_submissions/app.py [hunk 7]
```diff
@@ -77,42 +80,35 @@ def parse_s3_event_record(s3_event_record: dict) -> Tuple[str, str, int]:
...
-def parse_key(key: str) -> Tuple[Optional[str], Optional[str], Optional[str]]:
+def parse_key(key: str) -> Optional[dict]:
     """
     Parses the S3 object key to extract the collection UUID and curator tag, ignoring the REMOTE_DEV_PREFIX
+
+    Example of key with only curator_tag:
+    s3://<dataset submissions bucket>/<user_id>/<collection_id>/<curator_tag>
+
+    Example of key with dataset id:
+    s3://<dataset submissions bucket>/<user_id>/<collection_id>/<dataset_id>
```
Comment (danieljhegeman) [inline backend/corpora/dataset_submissions/app.py:91]:
```suggestion
    Example of key with only curator_tag:
    s3://<dataset submissions bucket>/<user_id>/<collection_id>/<curator_tag.h5ad>

    Example of key with dataset id:
    s3://<dataset submissions bucket>/<user_id>/<collection_id>/<dataset_id.h5ad>
```

File: backend/corpora/dataset_submissions/app.py (hunk 6 above)
Comment (Bento007) [reply backend/corpora/dataset_submissions/app.py:20]: I thought I fixed that. Good catch.

**Review by danieljhegeman — APPROVED** [review]
I think we're good!

**Pull request merged**